test/unit/transaction_test.js

```diff
@@ -489,131 +485,23 @@ describe('Transaction', function() {
-      expect(operation.destination).to.be.equal(destination);
-      expect(operation.amount).to.be.equal(amount);
-
-      const expectedXDR =
-        'AAAABQAAAADgSJG2GOUMy/H9lHyjYZOwyuyytH8y0wWaoc596L+bEgAAAAAAAADIAAAAAgAAAABzdv3ojkzWHMD7KUoXhrPx0GH18vHKV0ZfqpMiEblG1gAAAGQAAAAAAAAACAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAQAAAA9IYXBweSBiaXJ0aGRheSEAAAAAAQAAAAAAAAABAAAAAOBIkbYY5QzL8f2UfKNhk7DK7LK0fzLTBZqhzn3ov5sSAAAAAAAAAASoF8gAAAAAAAAAAAERuUbWAAAAQK933Dnt1pxXlsf1B5CYn81PLxeYsx+MiV9EGbMdUfEcdDWUySyIkdzJefjpR5ejdXVp/KXosGmNUQ+DrIBlzg0AAAAAAAAAAei/mxIAAABAijIIQpL6KlFefiL4FP8UWQktWEz4wFgGNSaXe7mZdVMuiREntehi1b7MRqZ1h+W+Y0y+Z2HtMunsilT2yS5mAA==';
-
-      expect(
-        transaction
-          .toEnvelope()
-          .toXDR()
-          .toString('base64')
-      ).to.be.equal(expectedXDR);
-      let expectedTxEnvelope = StellarBase.xdr.TransactionEnvelope.fromXDR(
-        expectedXDR,
-        'base64'
-      ).value();
-
-      expect(transaction.source).to.equal(
-        StellarBase.StrKey.encodeEd25519PublicKey(
-          expectedTxEnvelope
-            .tx()
-            .innerTx()
-            .value()
-            .tx()
-            .sourceAccount()
-            .ed25519()
-        )
-      );
-      expect(transaction.feeSource).to.equal(
-        StellarBase.StrKey.encodeEd25519PublicKey(
-          expectedTxEnvelope
-            .tx()
-            .feeSource()
-            .ed25519()
-        )
-      );
-
-      expect(transaction.innerFee).to.equal(
-        expectedTxEnvelope
-          .tx()
-          .innerTx()
-          .value()
-          .tx()
-          .fee()
-          .toString()
-      );
-      expect(transaction.fee).to.equal(
-        expectedTxEnvelope
-          .tx()
-          .fee()
-          .toString()
-      );
-
-      expect(transaction.innerSignatures.length).to.equal(1);
-      expect(
-        transaction.innerSignatures[0].toXDR().toString('base64')
-      ).to.equal(
-        expectedTxEnvelope
-          .tx()
-          .innerTx()
-          .value()
-          .signatures()[0]
-          .toXDR()
-          .toString('base64')
-      );
-
-      expect(transaction.signatures.length).to.equal(1);
-      expect(transaction.signatures[0].toXDR().toString('base64')).to.equal(
-        expectedTxEnvelope
-          .signatures()[0]
-          .toXDR()
-          .toString('base64')
-      );
-
-      let tampered = transaction.toEnvelope();
-      tampered._switch = {
-        name: 'envelopeTypeTxInvalid',
-        value: 99
-      };
-
-      expect(() => {
-        new StellarBase.Transaction(tampered, networkPassphrase);
-      }).to.throw(/Invalid EnvelopeType: envelopeTypeTxInvalid./);
-
-      transaction._envelopeType = {
-        name: 'envelopeTypeTxInvalid',
-        value: 99
-      };
-
-      expect(() => {
-        transaction.toEnvelope();
-      }).to.throw(/Invalid EnvelopeType: envelopeTypeTxInvalid./);
-
-      done();
+      expect(txWithMuxedAccount.source).to.equal(tx.source);
```

Comment:
why are they equal? isn't `txWithMuxedAccount.source` a muxed account while `tx.source` is not?

Reply:
source contains the strkey representation, my understanding is that for muxed accounts we'll show the public key not the `M` version

Reply:
I thought the strkey will print muxed accounts as "M...." instead of "G..."

Reply:
It does. What I mean is that we are showing the strkey representation of the ed25519 part of the muxed account, not the whole muxed account.

Reply:
what's the reasoning behind that? In horizon we convert muxed accounts to ed25519 account ids in ingestion so that we only ingest "G..." addresses. but I don't see why we should do the conversion for other cases

Reply:
@tamirms oh, my understanding is that we were going to show only `G..` for now.

Reply:
yeah we shouldn't show this in an SDK yet. It has no functionality and is confusing. We want this change to be transparent. If an SDK returns an M address all kinds of things downstream could potentially break.

Reply:
@tamirms makes some excellent points offline which I think guide the answer to this.

>for sep10 I think we want to ensure that no mux accounts are used in the source accounts. if we use the js sdk to convert a challenge string to an object we will not be able to verify that condition correctly

>     If an SDK returns an M address all kinds of things downstream could potentially break.

this is a valid concern and that's why horizon should only ingest G addresses. address fields in horizon responses will always contain G addresses . so the only case where you can encounter M addresses is when you're parsing xdr which actually contains an M address. If the user of the sdk created the xdr string that they are parsing then they're aware that it contains an M address. If the user of the sdk is parsing an xdr string generated by someone else it's unclear to me if there is a default behavior we should support. In many cases maybe the user doesn't care that it's an M address and they would have preferred the js sdk to convert automatically. But there might be other cases similar to sep10 where there are security implications to silently converting an M address to a G address

> alternatively, we could throw an exception if you try to parse an xdr which contains an M address

I think the exception idea could be a good one. It would mean that a caller would have to handle the theoretical possibility of receiving something other than a G, which seems like a good safety move.

Reply:
I fixed this, we now return the M or G address, also it uncovered a bug! https://github.com/stellar/js-stellar-base/commit/f050af83ba65cc305692a94bde6f679924836fcf